### ISSUE TITLE
CMake doesn't set up Python correctly despite specifying flags for multiple Python versions

### ISSUE THREAD

**Issue by sshkhr** [opencv/opencv#13662]
##### System information (version)

- OpenCV => 3.4.4
- Operating System / Platform => Ubuntu 16.04
- Compiler => gcc (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609

##### Detailed description

I am trying to install OpenCV3.4.4 with contrib modules. The issue is I have several python versions in my system and I'm unable to to get it to install in the required place even after correctly setting up all cmake flags.

##### Steps to reproduce

```
mkdir build
cd build
sudo cmake -D CMAKE_BUILD_TYPE=RELEASE \
                    -D CMAKE_INSTALL_PREFIX=/usr/local \ 
                    -D INSTALL_PYTHON_EXAMPLES=ON \ 
                    -D INSTALL_C_EXAMPLES=OFF \
                    -D OPENCV_ENABLE_NONFREE=ON \
                    -D OPENCV_EXTRA_MODULES_PATH=~/OpenCV/opencv_contrib/modules \
                    -D PYTHON_DEFAULT_EXECUTABLE=~/anaconda3/bin/python3.6 \ 
                    -D PYTHON_PACKAGES_PATH=~/anaconda3/lib/python3.6/site-packages/ \
                    -D PYTHON_LIBRARY=~/anaconda3/lib/libpython3.6m.so \
                    -D PYTHON_INCLUDE=~/anaconda3/include/python3.6m \
                    -D PYTHON_NUMPY_INCLUDE_DIRS=~/anaconda3/lib/python3.6/site -packages/numpy/core/include \ 
                    -D BUILD_NEW_PYTHON_SUPPORT=ON \ 
                    -D BUILD_opencv_python3=ON \
                    -D HAVE_opencv_python3=ON ..
```

The entire output is attached as a text file:

[op_log.txt](https://github.com/opencv/opencv/files/2777273/op_log.txt)

Of relevance:

--   Python 3:
--     Interpreter:                 /usr/bin/python3 (ver 3.5.2)
--     Libraries:                   NO
--     numpy:                       NO (Python3 wrappers can not be generated)
--     packages path:               lib/python3.5/dist-packages

--   Python (for build):            /home/sheshank/anaconda3/bin/python3.6

--   Install to:                    /usr/local

As you can see it is somehow installing in the default python3 instead of the ananconda3 python which I had specified. Please explain how I can install it in the anaconda python. Thanks in advance

**Comment by alalek** [comment]
> ~/

Avoid using of this bash extension for HOME directory. Try `$HOME/` instead.

**Comment by alalek** [comment]
Use `cmake --trace-expand ...` to debug it (from clean build directory).

**Comment by alalek** [comment]
> op_log.txt

Use clean build directory if you "playing" with CMake options. Some values can be cached by CMake scripts.

---

BTW, there is another issue:
> WARNING: Can't build ffmpeg test code

**Comment by sshkhr** [comment]
Thanks for the heads up with the ffmpeg issue. I am not working on videos currently so it won't be needed. I am using a clean build directory each time I run the cmake.
 
Could you help me out with what should be the value of the OPENCV_PYTHON3_INSTALL_PATH flag in cmake? I think it should be one of these:

```
-D OPENCV_PYTHON3_INSTALL_PATH=/home/sheshank/anaconda3/pkgs 
or 
-D OPENCV_PYTHON3_INSTALL_PATH=/home/sheshank/anaconda3/lib/python3.6/site-packages
```

I think fixing this should install opencv in the correct place

**Comment by abhiTronix** [comment]
@sshkhr why are you using `cmake` with `sudo`? That might be the problem.

**Comment by sshkhr** [comment]
I removed build anc CMake files folder and ran again without sudo (both cmake and make)

```
cmake -D CMAKE_BUILD_TYPE=RELEASE -D INSTALL_C_EXAMPLES=OFF -D OPENCV_ENABLE_NONFREE=ON -D OPENCV_EXTRA_MODULES_PATH=/home/sheshank/OpenCV/opencv_contrib/modules -D BUILD_opencv_python3=ON -D PYTHON3_EXECUTABLE=/home/sheshank/anaconda3/bin/python3.6 -D PYTHON3_PACKAGES_PATH=/home/sheshank/anaconda3/lib/python3.6/site-packages/ -D PYTHON3_LIBRARIES=/home/sheshank/anaconda3/lib/libpython3.6m.so -D PYTHON3_INCLUDE=/home/sheshank/anaconda3/include/python3.6m -D PYTHON3_NUMPY_INCLUDE_DIRS=/home/sheshank/anaconda3/lib/python3.6/site-packages/numpy/core/include -D PYTHON3_NUMPY_INCLUDE_DIRS=/home/sheshank/anaconda3/lib/python3.6/site-packages/numpy/core/include -D INSTALL_PYTHON_EXAMPLES=ON -D PYTHON_DEFAULT_EXECUTABLE=/home/sheshank/anaconda3/bin/python3.6 -D PYTHON_INCLUDE_DIR=/home/sheshank/anaconda3/include/python3.6m/ -D OPENCV_PYTHON3_INSTALL_PATH=/home/sheshank/anaconda3/pkgs -D WITH_FFMPEG=ON ..
```
I got the following error on running make:
```
/home/sheshank/OpenCV/opencv/3rdparty/protobuf/src/google/protobuf/text_format.cc:1309:1: error: prototype for ‘google::protobuf::TextFormat::Parser::Parser(bool)’ does not match any in class ‘google::protobuf::TextFormat::Parser’
 TextFormat::Parser::Parser(bool allow_unknown_field)
 ^
In file included from /home/sheshank/OpenCV/opencv/3rdparty/protobuf/src/google/protobuf/text_format.cc:43:0:
/home/sheshank/anaconda3/include/google/protobuf/text_format.h:458:28: error: candidates are: google::protobuf::TextFormat::Parser::Parser(const google::protobuf::TextFormat::Parser&)
   class LIBPROTOBUF_EXPORT Parser {
                            ^
/home/sheshank/anaconda3/include/google/protobuf/text_format.h:460:5: error:                 google::protobuf::TextFormat::Parser::Parser()
     Parser();
     ^
3rdparty/protobuf/CMakeFiles/libprotobuf.dir/build.make:1262: recipe for target '3rdparty/protobuf/CMakeFiles/libprotobuf.dir/src/google/protobuf/text_format.cc.o' failed
make[2]: *** [3rdparty/protobuf/CMakeFiles/libprotobuf.dir/src/google/protobuf/text_format.cc.o] Error 1
```

**Comment by alalek** [comment]
> opencv/3rdparty/protobuf/src
> anaconda3/include/google/protobuf

Headers and sources doesn't match. Extra include dirs were added (anaconda3/include) which breaks build - `make VERBOSE=1` shows complete compiler options.
To find source of problem inspect `CMakeVars.txt` and output of `cmake --trace-expand ...`

**Comment by Ayushmaniar** [comment]
> ##### System information (version)
> 
>     * OpenCV => 3.4.4
> 
>     * Operating System / Platform => Ubuntu 16.04
> 
>     * Compiler => gcc (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
> 
> 
> ##### Detailed description
> 
> I am trying to install OpenCV3.4.4 with contrib modules. The issue is I have several python versions in my system and I'm unable to to get it to install in the required place even after correctly setting up all cmake flags.
> ##### Steps to reproduce
> 
> ```
> mkdir build
> cd build
> sudo cmake -D CMAKE_BUILD_TYPE=RELEASE \
>                     -D CMAKE_INSTALL_PREFIX=/usr/local \ 
>                     -D INSTALL_PYTHON_EXAMPLES=ON \ 
>                     -D INSTALL_C_EXAMPLES=OFF \
>                     -D OPENCV_ENABLE_NONFREE=ON \
>                     -D OPENCV_EXTRA_MODULES_PATH=~/OpenCV/opencv_contrib/modules \
>                     -D PYTHON_DEFAULT_EXECUTABLE=~/anaconda3/bin/python3.6 \ 
>                     -D PYTHON_PACKAGES_PATH=~/anaconda3/lib/python3.6/site-packages/ \
>                     -D PYTHON_LIBRARY=~/anaconda3/lib/libpython3.6m.so \
>                     -D PYTHON_INCLUDE=~/anaconda3/include/python3.6m \
>                     -D PYTHON_NUMPY_INCLUDE_DIRS=~/anaconda3/lib/python3.6/site -packages/numpy/core/include \ 
>                     -D BUILD_NEW_PYTHON_SUPPORT=ON \ 
>                     -D BUILD_opencv_python3=ON \
>                     -D HAVE_opencv_python3=ON ..
> ```
> 
> The entire output is attached as a text file:
> 
> [op_log.txt](https://github.com/opencv/opencv/files/2777273/op_log.txt)
> 
> Of relevance:
> 
> -- Python 3:
> -- Interpreter: /usr/bin/python3 (ver 3.5.2)
> -- Libraries: NO
> -- numpy: NO (Python3 wrappers can not be generated)
> -- packages path: lib/python3.5/dist-packages
> 
> -- Python (for build): /home/sheshank/anaconda3/bin/python3.6
> 
> -- Install to: /usr/local
> 
> As you can see it is somehow installing in the default python3 instead of the ananconda3 python which I had specified. Please explain how I can install it in the anaconda python. Thanks in advance

Hey man, I see that you have struggled a lot with the python paths, I did too just a few minutes before. I would highly recommend you to use **cmake-gui** where you can manually type the paths. So remove all the paths that are automatically detected by cmake-gui for **python2** . Then click on configure and generate. 
Once that is done,check the python 3 in the output.

If that is good to go then go ahead and do make and then sudo make install.
Hope this helps 

PS : Dont forget to sym link the installed opencv library if you need to.
